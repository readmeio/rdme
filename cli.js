const path = require('path');
// We have to do this otherwise `require('config')` loads
// from the cwd where the user is running `rdme` which
// wont be what we want
//
// This is a little sketchy overwriting environment variables
// but since this is only supposed to be a cli and not
// requireable, i think this is okay
const configDir = process.env.NODE_CONFIG_DIR;
process.env.NODE_CONFIG_DIR = path.join(__dirname, '/config');

const config = require('config');

process.env.NODE_CONFIG_DIR = configDir;

const { version } = require('./package.json');
const configStore = require('./utils/configstore');

function load(command = '', subcommand = '') {
  const file = path.join(__dirname, 'lib', command, subcommand);
  try {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    return require(file);
  } catch (e) {
    const error = new Error('Command not found.');
    error.description = `Type ${`${config.cli} help`.yellow} to see all commands`;
    throw error;
  }
}

module.exports = function(cmd, args, opts = {}) {
  if (opts.version && (!cmd || cmd === 'help')) return Promise.resolve(version);

  let command = cmd || '';
  let subcommand;

  if (command.includes(':')) {
    [command, subcommand] = cmd.split(':');
  }

  const optsWithStoredKey = Object.assign({}, { key: configStore.get('apiKey') }, opts);

  try {
    return load(opts.help ? 'help' : command, subcommand).run({ args, opts: optsWithStoredKey });
  } catch (e) {
    return Promise.reject(e);
  }
};
