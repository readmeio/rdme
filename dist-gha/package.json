{
  "name": "rdme",
  "version": "9.0.0-yooo",
  "description": "ReadMe's official CLI and GitHub Action.",
  "private": true,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/readmeio/rdme.git"
  },
  "bugs": {
    "url": "https://github.com/readmeio/rdme/issues"
  },
  "type": "module",
  "oclif": {
    "bin": "rdme",
    "dirname": "rdme",
    "commands": {
      "strategy": "explicit",
      "target": "./cmds.js",
      "identifier": "COMMANDS"
    },
    "topicSeparator": ":",
    "hooks": {
      "prerun": {
        "target": "./cmds.js",
        "identifier": "prerun"
      },
      "createGHA": {
        "target": "./cmds.js",
        "identifier": "createGHA"
      }
    },
    "topics": {
      "categories": {
        "description": "List or create categories in your ReadMe developer hub."
      },
      "docs": {
        "description": "Sync or prune Guides pages in your ReadMe developer hub."
      },
      "openapi": {
        "description": "Manage your API definition. Supports OpenAPI, Swagger, Postman."
      },
      "versions": {
        "description": "Manage your documentation versions."
      }
    }
  }
}
