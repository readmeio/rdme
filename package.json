{
  "name": "rdme",
  "version": "7.2.0",
  "description": "ReadMe's official CLI and GitHub Action wrapper.",
  "license": "MIT",
  "author": "ReadMe <support@readme.io> (https://readme.com)",
  "engines": {
    "node": ">=14"
  },
  "bin": {
    "rdme": "bin/rdme"
  },
  "tags": [
    "api",
    "apis",
    "openapi",
    "swagger",
    "openapi initiative",
    "openapi specification",
    "openapi spec",
    "openapi document",
    "oai",
    "oas",
    "apidoc",
    "microservice",
    "documentation",
    "readme"
  ],
  "repository": {
    "type": "git",
    "url": "git://github.com/readmeio/rdme.git"
  },
  "bugs": {
    "url": "https://github.com/readmeio/rdme/issues"
  },
  "dependencies": {
    "@actions/core": "^1.6.0",
    "chalk": "^4.1.2",
    "cli-table": "^0.3.1",
    "command-line-args": "^5.2.0",
    "command-line-usage": "^6.0.2",
    "config": "^3.1.0",
    "configstore": "^5.0.0",
    "debug": "^4.3.3",
    "editor": "^1.0.0",
    "enquirer": "^2.3.0",
    "form-data": "^4.0.0",
    "gray-matter": "^4.0.1",
    "isemail": "^3.1.3",
    "node-fetch": "^2.6.1",
    "oas-normalize": "^6.0.0",
    "open": "^8.2.1",
    "parse-link-header": "^2.0.0",
    "read": "^1.0.7",
    "semver": "^7.0.0",
    "tmp-promise": "^3.0.2",
    "update-notifier": "^5.1.0"
  },
  "devDependencies": {
    "@readme/eslint-config": "^8.7.0",
    "@readme/oas-examples": "^5.1.1",
    "alex": "^10.0.0",
    "eslint": "^8.13.0",
    "jest": "^28.0.3",
    "nock": "^13.2.4",
    "prettier": "^2.6.2"
  },
  "scripts": {
    "lint": "eslint . bin/rdme bin/set-version-output",
    "lint-docs": "alex .",
    "pretest": "npm run lint && npm run lint-docs",
    "prettier": "prettier --list-different --write \"./**/**.js\"",
    "release": "npx conventional-changelog-cli -i CHANGELOG.md -s",
    "test": "jest --coverage"
  },
  "jest": {
    "coverageThreshold": {
      "global": {
        "branches": 80,
        "functions": 80,
        "lines": 90,
        "statements": 90
      }
    },
    "setupFiles": [
      "./__tests__/set-node-env"
    ],
    "testPathIgnorePatterns": [
      "./__tests__/get-api-nock",
      "./__tests__/set-node-env"
    ]
  },
  "prettier": "@readme/eslint-config/prettier"
}
