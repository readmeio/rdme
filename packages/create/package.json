{
  "name": "@readme/create",
  "version": "0.0.2",
  "description": "Sync your OpenAPI definition to ReadMe",
  "bin": "./bin/create.sh",
  "repository": {
    "type": "git",
    "url": "https://github.com/readmeio/rdme.git",
    "directory": "packages/create"
  },
  "keywords": [
    "api",
    "apis",
    "readme",
    "openapi",
    "swagger"
  ],
  "author": "ReadMe <support@readme.io> (https://readme.com)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/readmeio/rdme/issues"
  },
  "homepage": "https://github.com/readmeio/rdme#readme"
}
