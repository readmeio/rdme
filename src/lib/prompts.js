const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
