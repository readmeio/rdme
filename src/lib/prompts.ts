import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
