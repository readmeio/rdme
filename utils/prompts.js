exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
