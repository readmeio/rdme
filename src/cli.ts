#! /usr/bin/env node
import * as core from '@actions/core';
import chalk from 'chalk';
import updateNotifier from 'update-notifier-cjs';

import pkg from '../package.json';

import { isGHA } from './lib/isCI';

import rdme from '.';

updateNotifier({ pkg }).notify();

rdme(process.argv.slice(2))
  .then((msg: string) => {
    if (msg) {
      // eslint-disable-next-line no-console
      console.log(msg);
      if (isGHA()) {
        core.setOutput('rdme', msg);
      }
    }
    return process.exit(0);
  })
  .catch((err: Error) => {
    let message = `Yikes, something went wrong! Please try again and if the problem persists, get in touch with our support team at ${chalk.underline(
      'support@readme.io',
    )}.`;

    if (err.message) {
      message = err.message;
    }

    /**
     * If we're in a GitHub Actions environment, log errors with that formatting instead.
     *
     * @see {@link https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message}
     * @see {@link https://github.com/actions/toolkit/tree/main/packages/core#annotations}
     */
    if (isGHA()) {
      return core.setFailed(message);
    }

    // If this is a soft error then we should output the result as a regular log but exit the CLI
    // with an error status code.
    if (err.name === 'SoftError') {
      // eslint-disable-next-line no-console
      console.log(err.message);
    } else {
      // eslint-disable-next-line no-console
      console.error(chalk.red(`\n${message}\n`));
    }

    return process.exit(1);
  });
