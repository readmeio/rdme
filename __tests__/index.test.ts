import prompts from 'prompts';

import { version as pkgVersion } from '../package.json';
import cli from '../src';
import conf from '../src/lib/configstore';

import getAPIMock from './helpers/get-api-mock';
import { after, before } from './helpers/get-gha-setup';

describe('cli', () => {
  it('command not found', async () => {
    await expect(cli(['no-such-command'])).rejects.toThrow('Command not found');
  });

  describe('--version', () => {
    it('should return version from package.json', async () => {
      await expect(cli(['--version'])).resolves.toBe(pkgVersion);
    });

    it('should return version if the `-V` alias is supplied', async () => {
      await expect(cli(['-V'])).resolves.toBe(pkgVersion);
    });

    it('should return version from package.json for help command', async () => {
      await expect(cli(['help', '--version'])).resolves.toBe(pkgVersion);
    });

    // This is necessary because we use --version for other commands like `docs`
    it('should only return version if no command', async () => {
      await expect(cli(['no-such-command', '--version'])).rejects.toThrow(
        // This can be ignored as it's just going to be a command not found error
        'Command not found.'
      );
    });
  });

  describe('--help', () => {
    beforeEach(() => {
      // The `table-layout` dependency within `command-line-usage` has a quirk in the 1.x family where if it's being
      // run within a Jest environment it won't always use the same amount of columns for tables, resulting in cases
      // where Jest snapshots will randomly return back different results.
      //
      // This issue has been fixed in later versions of `table-layout` but `command-line-usage` is still targeting the
      // 1.x series.
      //
      // See https://github.com/75lb/table-layout/issues/9 for more information.
      process.stdout.columns = 100;
      process.stderr.columns = 100;
    });

    /* eslint-disable @typescript-eslint/ban-ts-comment */
    afterEach(() => {
      // @ts-ignore This is the only way we can disable columns within the `table-layout` library.
      process.stdout.columns = undefined;
      // @ts-ignore
      process.stderr.columns = undefined;
    });
    /* eslint-enable @typescript-eslint/ban-ts-comment */

    it('should print help', async () => {
      await expect(cli(['--help'])).resolves.toContain('a utility for interacting with ReadMe');
    });

    it('should print help for the `-H` alias', async () => {
      await expect(cli(['-H'])).resolves.toContain('a utility for interacting with ReadMe');
    });

    it('should print usage for a given command', async () => {
      await expect(cli(['openapi', '--help'])).resolves.toMatchSnapshot();
    });

    it('should print usage for a given command if supplied as `help <command>`', async () => {
      await expect(cli(['help', 'openapi'])).resolves.toMatchSnapshot();
    });

    it('should not surface args that are designated as hidden', async () => {
      await expect(cli(['openapi', '--help'])).resolves.toMatchSnapshot();
    });

    it('should show related commands for a subcommands help menu', async () => {
      await expect(cli(['versions', '--help'])).resolves.toMatchSnapshot();
    });
  });

  describe('subcommands', () => {
    it('should load subcommands from the folder', async () => {
      await expect(cli(['openapi:validate', 'package.json'])).rejects.not.toThrow('Command not found.');
    });
  });

  it('should not error with undefined cmd', async () => {
    await expect(cli([])).resolves.toContain('Upload OpenAPI/Swagger definitions');
  });

  it('should add stored apiKey to opts', async () => {
    expect.assertions(1);
    const key = '123456';
    const version = '1.0.0';
    conf.set('apiKey', key);

    const versionMock = getAPIMock().get(`/api/v1/version/${version}`).basicAuth({ user: key }).reply(200, { version });

    await expect(cli(['docs', `--version=${version}`])).rejects.toStrictEqual(
      new Error('No path provided. Usage `rdme docs <path> [options]`.')
    );

    conf.clear();
    versionMock.done();
  });

  it('should error with `rdme oas` arguments passed in', async () => {
    await expect(cli(['oas', 'endpoint'])).rejects.toThrow(/.*/);
  });

  describe('GHA onboarding via @supportsGHA decorator', () => {
    let consoleInfoSpy;
    const key = '123';
    const slug = 'new-doc';

    beforeEach(() => {
      before(() => true);
      consoleInfoSpy = jest.spyOn(console, 'info').mockImplementation();
    });

    afterEach(() => {
      after();
      consoleInfoSpy.mockRestore();
    });

    const getCommandOutput = () => {
      return [consoleInfoSpy.mock.calls.join('\n\n')].filter(Boolean).join('\n\n');
    };

    it.each([
      ['changelogs', 'changelogs', ''],
      ['custompages', 'custompages', ''],
    ])('should run GHA workflow for the %s command', async (cmd, type, file) => {
      expect.assertions(3);
      prompts.inject([false]);

      const getMock = getAPIMock().get(`/api/v1/${type}/${slug}`).basicAuth({ user: key }).reply(404, {});

      await expect(
        cli([cmd, '--dryRun', `--key=${key}`, `__tests__/__fixtures__/${type}/new-docs/${file}`])
      ).rejects.toStrictEqual(
        new Error(
          'GitHub Actions workflow creation cancelled. If you ever change your mind, you can run this command again with the `--github` flag.'
        )
      );

      const output = getCommandOutput();
      expect(output).toMatch(`dry run! This will create '${slug}'`);
      expect(output).toMatch("Looks like you're running this command in a GitHub Repository!");

      getMock.done();
    });
  });
});
