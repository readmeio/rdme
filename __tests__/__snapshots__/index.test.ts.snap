// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli --help should not surface args that are designated as hidden 1`] = `
"
Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

Usage

  rdme openapi [file] [options] 

Options

  --key string                Project API key                                                       
  --id string                 Unique identifier for your API definition. Use this if you're re-     
                              uploading an existing API definition                                  
  --version string            Project version                                                       
  --workingDirectory string   Working directory (for usage with relative external references)       
  -h, --help                  Display this usage guide                                              

Related commands

  $ rdme validate   Validate your OpenAPI/Swagger definition. 
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]    
"
`;

exports[`cli --help should print usage for a given command 1`] = `
"
Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

Usage

  rdme openapi [file] [options] 

Options

  --key string                Project API key                                                       
  --id string                 Unique identifier for your API definition. Use this if you're re-     
                              uploading an existing API definition                                  
  --version string            Project version                                                       
  --workingDirectory string   Working directory (for usage with relative external references)       
  -h, --help                  Display this usage guide                                              

Related commands

  $ rdme validate   Validate your OpenAPI/Swagger definition. 
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]    
"
`;

exports[`cli --help should print usage for a given command if supplied as \`help <command>\` 1`] = `
"
Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

Usage

  rdme openapi [file] [options] 

Options

  --key string                Project API key                                                       
  --id string                 Unique identifier for your API definition. Use this if you're re-     
                              uploading an existing API definition                                  
  --version string            Project version                                                       
  --workingDirectory string   Working directory (for usage with relative external references)       
  -h, --help                  Display this usage guide                                              

Related commands

  $ rdme validate   Validate your OpenAPI/Swagger definition. 
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]    
"
`;

exports[`cli --help should show related commands for a subcommands help menu 1`] = `
"
List versions available in your project or get a version by SemVer (https://semver.org/).

Usage

  rdme versions [options] 

Options

  --key string       Project API key                                                 
  --version string   A specific project version to view                              
  --raw              Return raw output from the API instead of in a \\"pretty\\" format. 
  -h, --help         Display this usage guide                                        

Related commands

  $ rdme versions:create   Create a new version for your project.                
  $ rdme versions:update   Update an existing version for your project.          
  $ rdme versions:delete   Delete a version associated with your ReadMe project. 
"
`;
